src/Compilers/CSharp/Portable/Syntax/NonRecursiveSyntaxRewriter.cs

```diff
@@ -0,0 +1,286 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Represents a non-recursive visitor which descends an entire <see cref="CSharpSyntaxNode"/> graph and
+    /// may replace or remove visited SyntaxNodes in depth-first order.
+    /// </summary>
+    public partial class NonRecursiveSyntaxRewriter : CSharpSyntaxVisitor<NonRecursiveSyntaxRewriter.Chunk>
+    {
+        public class Chunk
```

Comment:
>Chunk [](start = 21, length = 5)

I'd like this a lot better if it could be a struct, not causing allocation.   And instead of a captured transform function, it could be described as data, or maybe inferred by the information you already have.  For example, you already know if you have a token and need to call TransformToken.

I'm also concern that this type is public and passed through the generated visitor API.